simulator/resourcewatcher/resourcewatcher.go

```diff
@@ -0,0 +1,323 @@
+// and knows where to send the event.
+type resourceEventProxy struct {
+	// writer knows where to send the event.
+	writer StreamWriter
+	// The RESTClient to watch the specified.
+	c cache.Getter
+	// The kind of resource to watch.
+	r resourceKind
+	// The Object of resource to watch.
+	o runtime.Object
+	// The last value of ResourceVersion. This is used to RetryWatcher.
+	// First, this value is specified by a user.
+	// After that this is updated when received an event everytime.
+	//
+	// The RetryWatcher will be reconnect to the apiserver if it is disconnected.
+	// lastResourceVersion can be used to ensure that only events
+	// that have not yet been received are received when reconnecting.
+	lastResourceVersion string
+}
+
+func newresourceEventProxy(sw StreamWriter, c cache.Getter, r resourceKind, o runtime.Object, ilrv string) *resourceEventProxy {
+	return &resourceEventProxy{
+		writer:              sw,
+		c:                   c,
+		r:                   r,
+		o:                   o,
+		lastResourceVersion: ilrv,
+	}
+}
+
+// ListAndHandleItems calls the list for the resource and the results is send to the client by ListItemsHandler method.
+func (p *resourceEventProxy) ListAndHandleItems(lw cache.Lister) error {
+	list, err := lw.List(metav1.ListOptions{})
+	if err != nil {
+		return xerrors.Errorf("failed to list: %w", err)
+	}
+	items, lrv, err := extractListItem(list)
+	if err != nil {
+		return xerrors.Errorf("call HandleListItems: %w", err)
+	}
+	if err := p.listItemsHandler(items); err != nil {
+		return xerrors.Errorf("call ListItemsHandle: %w", err)
+	}
+	p.lastResourceVersion = lrv
+	return nil
+}
+
+// extractListItem validates whether the object is a list object and returns these items and lastResourceVersion.
+func extractListItem(list runtime.Object) ([]runtime.Object, string, error) {
+	listMetaInterface, err := meta.ListAccessor(list)
+	if err != nil {
+		return nil, "", xerrors.Errorf("failed to ListAccessor, unable to understand list result %#v: %w", list, err)
+	}
+	items, err := meta.ExtractList(list)
+	if err != nil {
+		return nil, "", xerrors.Errorf("failed to ExtractList, unable to understand list result %#v: %w", list, err)
+	}
+	return items, listMetaInterface.GetResourceVersion(), nil
+}
+
+// WatchAndHandleEvent prepares a handler for the wacher and runs the handler
+// until the stopCh is closed.
+func (p *resourceEventProxy) WatchAndHandleEvent(watcher watch.Interface, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	handleFunc := p.watchHandlerFunc(watcher)
+	run := func(stopCh <-chan struct{}) {
+		if err := handleFunc(stopCh); err != nil {
+			p.watchErrorHandler(err)
+		}
+	}
+	var wg wait.Group
+	wg.StartWithChannel(stopCh, run)
+	wg.Wait()
+}
+
+// listItemsHandler sends results of list as "ADDED" event to the client.
+// This method will be expected to call before starting the watch.
+func (p *resourceEventProxy) listItemsHandler(items []runtime.Object) error {
```

Comment:
Let's rename this to sendListedItems. `listItemsHandler ` doesn't tell how `listItems` will be handled by this method.